### PR TITLE
ch4: fix partioned communication status check

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

There is a possibility the next round of CTS can arrive before the sender side completes. This bug sporadically triggered and happens more often on FreeBSD. This PR fixes it by always using arithmetic rather than resetting the status on completion. (`MPL_atomic_fetch_sub_int` rather than `MPL_atomic_store_int`).

reference current nightly tests https://jenkins-pmrs.cels.anl.gov/job/mpich-main-ch4-freebsd/229/testReport/
```
summary_junit_xml. - ./part/start_pready 2 -spart=1 -rpart=1 -tot_count=64 -range=1 -iteration=5 | 3 min 0 sec | 1
summary_junit_xml. - ./part/parrived_wait 2 -spart=1 -rpart=1 -tot_count=64 -iteration=5 | 3 min 0 sec | 1
```
[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/mpid/ch4/generic/am/mpidig_am_part_utils.h [hunk 1]
```diff
@@ -9,20 +9,35 @@
...
+ *     1. Precv_init:    status = 0
+ *     2. receive MPIDIG_PART_SEND_INIT: status +=1
+ *     3. match and send CTS
+ *     4. Start (local): status += 1
+ *     5. receive all data and complete: status -= 1
+ *     ready for next start.
+ *
+ *     Note that 2 and 4 can take place in any order, and 3 takes place after 2 or 4, whichever is later.
+ *     After first round, it no longer need matching, so no longer need step 1, 2, and 3.
+ */
```
Comment (minsii) [inline src/mpid/ch4/generic/am/mpidig_am_part_utils.h:33]: In the second or later round, receiver still needs to send CTS after locally started.

File: src/mpid/ch4/generic/am/mpidig_am_part.c [hunk 2]
```diff
@@ -55,7 +55,7 @@ static int part_req_create(void *buf, int partitions, MPI_Aint count,
 static void part_req_am_init(MPIR_Request * part_req)
 {
     MPIDIG_PART_REQUEST(part_req, peer_req_ptr) = NULL;
-    MPL_atomic_store_int(&MPIDIG_PART_REQUEST(part_req, status), MPIDIG_PART_REQ_UNSET);
+    MPL_atomic_store_int(&MPIDIG_PART_REQUEST(part_req, status), 0);
 }
```
Comment (minsii) [inline src/mpid/ch4/generic/am/mpidig_am_part.c:59]: Can we keep using predefined MACRO instead of magic numbers?

File: src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c [hunk 3]
```diff
@@ -144,8 +144,11 @@ int MPIDIG_part_send_data_target_msg_cb(int handler_id, void *am_hdr, void *data
     MPIR_Request *part_rreq = msg_hdr->rreq_ptr;
     MPIR_Assert(part_rreq);
 
-    /* Erroneous if received data with a non-CTS rreq */
-    MPIDIG_PART_CHECK_RREQ_CTS(part_rreq);
+#ifdef HAVE_ERROR_CHECKING
+    /* If this happens, fix the code */
+    MPIR_Assert(MPL_atomic_load_int(&MPIDIG_PART_REQUEST(part_rreq, status)) ==
+                MPIDIG_PART_STATUS_CAN_CTS);
+#endif
```
Comment (minsii) [inline src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:152]: The error handling change itself is OK, but better to move to a separate commit.

File: src/mpid/ch4/generic/am/mpidig_am_part_utils.h [hunk 4]
```diff
@@ -9,20 +9,35 @@
...
-#define MPIDIG_PART_REQ_CTS 2   /* Indicating receiver is ready to receive data
-                                 * (requests matched and start called) */
-
-/* Status to deactivate at request completion */
-#define MPIDIG_PART_SREQ_INACTIVE 0     /* Sender need 2 increments to be CTS: start and remote start */
-#define MPIDIG_PART_RREQ_INACTIVE 1     /* Receiver need 1 increments to be CTS: start */
+
+#define MPIDIG_PART_STATUS_UNSET 0      /* Initial status set at psend|precv_init */
+#define MPIDIG_PART_STATUS_CAN_SEND 2   /* sender side started and got CTS, can send data */
+#define MPIDIG_PART_STATUS_CAN_CTS  2   /* receiver side started and matched, can send CTS */
```
Comment (minsii) [inline src/mpid/ch4/generic/am/mpidig_am_part_utils.h:41]: Can we embed sender-side and receiver-side hint into the second and the third statuses respectively? I.e., `MPIDIG_PART_STATUS_CAN_SEND` is used only on the sender side. In addition, `MPIDIG_PART_STATUS_UNSET` is unused after the change, but I'd recommend we use it instead of magic number (see above comment).

Additional macros might be added to express the -`1` and -`2` numbers when completing a request.

File: src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c [hunk 5]
```diff
@@ -31,7 +31,7 @@ static int part_send_data_target_cmpl_cb(MPIR_Request * rreq)
     MPID_Request_complete(rreq);
 
     /* Reset part_rreq status to inactive */
-    MPL_atomic_store_int(&MPIDIG_PART_REQUEST(part_rreq, status), MPIDIG_PART_RREQ_INACTIVE);
+    MPL_atomic_fetch_sub_int(&MPIDIG_PART_REQUEST(part_rreq, status), 1);
```
Comment (minsii) [inline src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: I'd suggest aways use macro instead of magic numbers. I also put the comment in the next commit.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: I disagree here. The semantics with the "status" here is a numeric semantics -- it starts as `0`, add `1` for each necessary condition is met; it checks agast a final value to decide whether "all" conditions are met. In this semantics, `0` is not magic. `0` means `0` as every numeric context everywhere. On the other hand, the names made the simple numeric context cryptic. For example, `UNSET + 1 -> PARTIAL_ACTIVATED`. `PARTIAL_ACTIVATED + 1 -> REQ_CTS`. It makes `0 + 1 = 1` and `1 + 1 = 2` obscure. In fact, I am leaning toward removing all names. Code like
```
if (status == 2) {
    /* Both conditions met, can send data */
    issue_data
}
```
seems very straight forward. On the other hand,
```
if (status == MPIDIG_PART_REQ_CTS) {
    issue_data
}
```
seems mysterious.

File: test/mpi/part/parrived.c [hunk 6]
```diff
@@ -87,7 +87,7 @@ int main(int argc, char *argv[])
     /* Send buffer size and receive buffer size must be identical */
     scount = tot_count / spart;
     rcount = tot_count / rpart;
-    if (spart * scount != rpart * rcount) {
+    if (spart * scount != tot_count || rpart * rcount != tot_count) {
         fprintf(stderr, "Invalid partitions (%d, %d) or tot_count (%ld),"
```
[the remark this answers is not in the document]
Comment (hzhou) [reply test/mpi/part/parrived.c:91]: Doesn't if make the code odd that if we ask it to send `1026` total count, but the code actually send `1024` count although the receive happens to make the same mistake?

[the remark this answers is not in the document]
Comment (minsii) [reply test/mpi/part/parrived.c:91]: I think such a case is fine as long as the actual numbers used by send/recv match.

File: src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c (hunk 5 above)
Comment (minsii) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: I guess that we cannot converge. Maybe you can ping the other reviewer to co-approve this commit?

Comment (raffenet) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: So when I was thinking about how to implement a CTS scheme in the netmod, I initially drew up local and remote epoch counters that would be incremented for each `MPI_Start`, and the receiver process could potentially use a remote atomic to update the remote epoch on the sender process. Data transfer would only start when `local == remote`. From what I can tell, this code implements essentially the same just with a single var, which does make it a little hard to understand for the reader. Both the old (buggy) version and this fixed version.

IMO, this version is better but still a bit difficult to understand. What do you think of having separate local and remote values?

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: I think it makes sense. But probably a better name/prefix is `sender_` and `receiver_` rather than `local_` and `remote_`.  How about let me drop the last commit, so @raffenet you could eventually replace it with your new versions.

Comment (raffenet) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: Sure, I can give that a try.

File: test/mpi/part/parrived.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply test/mpi/part/parrived.c:91]: It doesn't make sense to me, but I am removing this commit.

**Comment by hzhou** [comment]
test:mpich/ch4/most

File: src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c (hunk 5 above)
Comment (minsii) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: @hzhou can we replace direct use of `1` or `2`  with `(MPIDIG_PART_REQ_CTS-MPIDIG_PART_SREQ_INACTIVE)` or `(MPIDIG_PART_REQ_CTS-MPIDIG_PART_RREQ_INACTIVE)`? Without the additional comment in your other commit, the `1` and `2` are magic. If we want to keep this fix PR simple, I'd suggest we use the current MACROs.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: What does *subtract* `MPIDIG_PART_REQ_CTS-MPIDIG_PART_SREQ_INACTIVE` mean? It doesn't make sense to me.

Comment (minsii) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: It reads as reduce the status from CTS to INACTIVE.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: No. `a - 1` is not reduce `a` to `1`.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: @minsii I guess you haven't really understood what the bug was. Your assumption that `status` is `CTS` (2) is incorrect, and that was the bug. The `status` could be `3` due to the next `CTS` arrive before the previous send completes.

Comment (raffenet) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: Right. This is why I think having separate send and recv epoch vars would be easier to understand.

Comment (raffenet) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: To be clear, I think the fix in this PR is fine to merge for now.

Comment (minsii) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: I see the key issue here is that, the "remote-started" condition is updated by the remote process, thus it may be out-of-order with the update of the local conditions. The current code covers the cases of (1) local-completed -> local-started -> remote-started and (2) local-completed -> remote-started -> local-started. But it does not cover (3)  remote-started -> local-completed -> local-started. The (3) case is fixed in this PR. I think there should be no other cases.

I agreed that having separate local and remote vars  (or send/recv) gives a better reading. But I guess the same out-of-order issue may occur, as the remote var is updated by am.

Comment (minsii) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: The key concern of the second commit is that, the cntr/stat logic is a bit complex (it was a fault of my original code) thus we want to keep the logic definition in one place. The original code defines status macros in a single place as the logic definition. But now it splits into two parts -- the part uses the status, and the part directly uses -1 or -2 numbers.

I hope the logic can be still defined in a single place, thus don't make the code more complex. That was why suggested to still use status macros instead of -1/-2. Using `(CTS-INACTIVE)` might be misleading. How about define `NUM_CTS_CONDITIONS`? Besides, these `INACTIVE` status macros should be removed as they are unused now.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:35]: Alright, I'll name them.

Comment (hzhou) [reply src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c:36]: @minsii I added the last commit to address your concern.

**Review by minsii — APPROVED** [review]

**Review by raffenet — APPROVED** [review]
